iceoryx_hoofs/include/iceoryx_hoofs/internal/containers/uninitialized_array.hpp

```diff
@@ -0,0 +1,117 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_HOOFS_CONTAINERS_UNINITIALIZED_ARRAY_HPP
+#define IOX_HOOFS_CONTAINERS_UNINITIALIZED_ARRAY_HPP
+
+#include "iceoryx_hoofs/iceoryx_hoofs_types.hpp"
+
+#include <cstdint>
+
+namespace iox
+{
+namespace containers
+{
+/// @brief struct used as policy parameter in UninitializedArray to wrap an array with all elements zeroed
+/// @tparam ElementType is the array type
+/// @tparam Capacity is the array size
+template <typename ElementType, uint64_t Capacity>
+struct ZeroedBuffer
+{
+    // AXIVION Next Construct AutosarC++19_03-A18.1.1 : required by low level UninitializedArray building block and encapsulated in abstraction
+    // NOLINTBEGIN(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
+    using element_t = cxx::byte_t[sizeof(ElementType)];
+    // AXIVION Next Construct AutosarC++19_03-A18.1.1 : required by low level UninitializedArray building block and encapsulated in abstraction
+    alignas(ElementType) element_t value[Capacity]{{0}};
+    // NOLINTEND(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
+};
+
+/// @brief struct used as policy parameter in UninitializedArray to wrap an uninitialized array
+/// @tparam ElementType is the array type
+/// @tparam Capacity is the array size
+template <typename ElementType, uint64_t Capacity>
+struct NotZeroedBuffer
+{
+    // AXIVION Next Construct AutosarC++19_03-A18.1.1 : required by low level UninitializedArray building block and encapsulated in abstraction
+    // NOLINTBEGIN(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
+    using element_t = cxx::byte_t[sizeof(ElementType)];
+    // AXIVION Next Construct AutosarC++19_03-A18.1.1 : required by low level UninitializedArray building block and encapsulated in abstraction
+    alignas(ElementType) element_t value[Capacity];
+    // NOLINTEND(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
+};
+
+/// @brief Wrapper class for a C-style array of type ElementType and size Capacity. Per default it is uninitialized but
+/// all elements can be zeroed via template parameter ZeroedBuffer.
+/// @tparam ElementType is the array type
+/// @tparam Capacity is the array size
+/// @tparam Buffer is the policy parameter to choose between an uninitialized, not zeroed array (=NotZeroedBuffer,
+/// default) and an uninitialized array with all elements zeroed (=ZeroedBuffer)
+/// @note Out of bounds access leads to undefined behavior
+template <typename ElementType, uint64_t Capacity, template <typename, uint64_t> class Buffer = NotZeroedBuffer>
+class UninitializedArray
+{
+    static_assert(Capacity > 0U, "The size of the UninitializedArray must be greater than 0!");
```

Comment:
I created an issue to also support uninitialized arrays with a capacity of zero: https://github.com/eclipse-iceoryx/iceoryx/issues/1760

But this is some low prio task for the future.

Reply:
Use cases for this are rare (some meta programming tasks maybe ...). For now the static assert will do nicely.